libs/datamodel/core/src/transform/ast_to_dml/validate/referential_actions.rs

```diff
@@ -1,63 +1,148 @@
-use std::{collections::HashSet, fmt, rc::Rc};
-
-use dml::{datamodel::Datamodel, field::RelationField, model::Model, traits::WithName};
+mod visited_relation;
 
 use crate::{
     ast::Span,
     diagnostics::{DatamodelError, Diagnostics},
     transform::ast_to_dml::validate::RELATION_ATTRIBUTE_NAME,
 };
+use dml::{datamodel::Datamodel, field::RelationField, model::Model, traits::WithName};
+use itertools::Itertools;
+use std::{
+    collections::{BTreeSet, HashMap, HashSet},
+    rc::Rc,
+};
+use visited_relation::VisitedRelation;
 
-/// A linked list structure for visited relation paths.
-#[derive(Debug)]
-struct VisitedRelation<'a> {
-    previous: Option<Rc<VisitedRelation<'a>>>,
-    model_name: &'a str,
-    field_name: &'a str,
-}
+/// Find relations from the given model that end up into the same model,
+/// triggering cascading referential actions.
+pub(crate) fn detect_multiple_cascading_paths(
+    datamodel: &Datamodel,
+    parent_model: &Model,
+    parent_field: &RelationField,
+    span: Span,
+    errors: &mut Diagnostics,
+) {
+    // A graph of relations to traverse.
+    let mut next_relations = Vec::new();
+
+    // Gather all paths from this model to any other model, skipping
+    // cyclical relations
+    let mut paths = Vec::new();
+
+    // Add all relations from current model to the graph vector. At this point
+    // we only care about multiple paths from this model to any other model. We
+    // handle paths that cross, but are not started from here, when calling the
+    // function from corresponding models.
+    for field in parent_model.relation_fields().filter(|field| field.is_singular()) {
+        next_relations.push((
+            parent_model,
+            field,
+            Rc::new(VisitedRelation::root(parent_model.name(), field.name())),
+        ));
+    }
+
+    while let Some((model, field, visited_relations)) = next_relations.pop() {
+        let related_field = datamodel.find_related_field_bang(field).1;
+        let related_model = datamodel.find_model(&field.relation_info.to).unwrap();
+
+        // Following a directed graph, we don't need to go through back-relations.
+        if field.is_list() {
```

Comment:
For 1:1 relations, the backrelation is singular. I don't know if that throws a wrench in path detection in that case.